File: LTSts/Section332/Figure10.m
%%%% Simulation study of Section 3.3.2.

% Import data 
data = readtable("P87_KZ.csv");
Y = data.Y;
clear data;

n = 121; % number of data to use as training set
nfore = size(Y,1)-n;

reps = 10; % number of repetitions

model=struct;
model.s=54;
model.trend=1;
model.seasonal=101;
model.lshift=-1;

%% Without contamination

train = Y(1:n);

% Estimate
outEST=LTSts(train,'model',model,'plots',0);

% Forecast
outFORE = forecastTS(outEST,'model',model,'plots',1);
hold on;
plot(Y);
plot(outEST.outliers,train(outEST.outliers),'rx','LineWidth',1)    % outliers 
hold off;

%% With contamination

% Generation of random locations and signs of the contamination
rng(2023)
locations_noise = randi(n,reps,1);
sign_noise=randi(2,reps,1)-1;
sign_noise(~sign_noise)=-1;

for i=1:reps
    rng(1)

    train = Y(1:n);
    train(locations_noise(i)) = max( 0 , train(locations_noise(i)) + 20000*sign_noise(i) );
    
    % Estimate
    outEST=LTSts(train,'model',model,'plots',0);
    
    % Forecast
    outFORE = forecastTS(outEST,'model',model,'plots',1);
    hold on;
    plot(Y);
    plot(outEST.outliers,train(outEST.outliers),'rx','LineWidth',1)    % outliers 
    hold off;

end

File: LTSts/Section332/Figure10.R
library("seasonal") 
library("forecast")

# Import data
data <- read.csv("P87_KZ.csv",header=T)
Y <- data$Y
rm(data)

n <- 121 # number of data to use as training set

reps <- 10 # number of repetitions

y_full <- ts(data=Y,start=2010,frequency=12)

#### Without contamination ####

train <- Y[1:n]

x11()
plot(y_full,col='blue',ylim=c(0,500000),ylab='',lwd=2)

y_train <- ts(data=train,start=2010,frequency=12)
out <- seas(y_train,forecast.save="forecasts")
fore <- series(out, c("forecast.forecasts", "s12"))

lines(fore[,1],col='black')
lines(fore[,2],col="red")
lines(fore[,3],col="red")
lines(fore[,4],col="red")


#### With contamination ####

# random locations and signs of the contamination generated in Matlab (see code Figure10.m)
locations_noise <- c(39,108,72,16,18,57,3,89,64,66)
sign_noise <- c(-1,1,-1,-1,-1,-1,-1,-1,-1,-1)

for(i in 1:reps){
  
  train <- Y[1:n]
  train[locations_noise[i]] <- max( 0 , train[locations_noise[i]]+20000*sign_noise[i] )

  yi <- ts(data=train,start=2010,frequency=12)
  out <- seas(yi,forecast.save="forecasts")
  fore <- series(out, c("forecast.forecasts", "s12"))

  x11()
  plot(y_full,col='blue',ylim=c(0,500000),ylab='',lwd=2)
  lines(yi,col="deepskyblue",lwd=2);lines(y_train,col="blue",lwd=2)
  lines(fore[,1],col='black')
  lines(fore[,2],col="red")
  lines(fore[,3],col="red")
  lines(fore[,4],col="red")

}

